src/poll.c

```diff
@@ -0,0 +1,372 @@
+// poll locks
+
+// Lock poll data structure (poll_fd_array, poll_data_array, poll_num and 
+// poll_first_deprecated.
+pthread_mutex_t poll_array_mtx = PTHREAD_MUTEX_INITIALIZER;
+
+static inline int poll_array_lock() {
+	if(pthread_mutex_lock(&poll_array_mtx) == 0) {
+		return 0;
+	} else {
+		error("Unable to lock poll file list");
+		return -1;
+	}
+}
+
+static inline int poll_array_unlock() {
+	if(pthread_mutex_unlock(&poll_array_mtx) == 0) {
+		return 0;
+	} else {
+		error("Unable to unlock poll file list");
+		return -1;
+	}
+}
+
+// ---------------------------------------------------------------------------
+// internal signal handler used to interrupt poll.
+void poll_handler(int signo)
+{
+	if(signo){} // Remove compiler warning
+}
+
+// ---------------------------------------------------------------------------
+// internal methods
+int poll_time_update_nolock(struct timeval *tv) {
+	int retv = gettimeofday(tv, NULL);
+	if(retv != 0) {
+		error("Could not get current time");
+	}
+	return retv;
+}
+
+// Updates p->tv to current time.
+// This is not thread save.
+// concurrent reads are save. Concurrent writes need to be syncronized.
+int poll_data_time_update_readsave_nolock(struct poll_data *d) {
+	static struct timeval *buff;
+	if(!buff) {
+		buff = malloc(sizeof(struct timeval));
+		if(!buff) {
+			error("Could not allocate buffer for poll_data_time_update_readsave_nolock()");
+			return -1;
+		}
+	}
+
+	buff->tv_sec = d->tv->tv_sec;
+	buff->tv_usec = d->tv->tv_sec;
+	int retv = poll_time_update_nolock(buff);
+
+	// Switch pointers.
+	struct timeval *tv = d->tv;
+	d->tv = buff;
+	buff = tv;
+
+	return retv;
+}
+
+// if bigger is less than lower we return 0.
+unsigned long long poll_time_difference_nolock(struct timeval *bigger, struct timeval *lower) {
+	struct timeval diff;
+	if(timercmp(bigger, lower, >)) {
+		timersub(bigger, lower, &diff);
+		return (diff.tv_sec * 1000000) + diff.tv_usec;
+	} else {
+		return 0;
+	}
+}
+
+int poll_array_search_nolock(char *path, int type) {
+	uint32_t hash = simple_hash(path);
+	int i;
+	for(i = 0; i < poll_num; i++)
+		if(poll_fd_array[i].fd < 0 && poll_data_array[i].path_hash == hash && poll_fd_array[i].events == type && \
+				strncmp(poll_data_array[i].path, path, strlen(poll_data_array[i].path)) == 0)
+			return i;
+	return -1;
+}
+
+int poll_array_find_next_deprecated_nolock(int index) {
+	for(index++; index<poll_num; index++)
+		if(poll_fd_array[index].fd < 0)
+			return index;
+	return -1;
+}
+
+void poll_array_remove_nolock(int index) {
+	if(poll_fd_array[index].fd >= 0)
+		if(close(poll_fd_array[index].fd) != 0)
+			error("Failed to proper close file descriptor %d", poll_fd_array[index].fd);
+	poll_fd_array[index].fd = -1;	
+	if(poll_first_deprecated < index)
+		poll_first_deprecated = index;
+}
+
+int poll_array_add_nolock(char *path, int type) {
+	int index;
+
+	if(poll_first_deprecated >= 0) {
+		index = poll_first_deprecated;
```

Comment:
The line above: `poll_first_deprecated >= 0`.
If `poll_first_deprecated == -1` we do have to increase the array.